saleor/order/fetch.py

```diff
@@ -145,22 +182,63 @@ def fetch_draft_order_lines_info(
                 voucher=None,
                 voucher_code=None,
                 voucher_denormalized_info=None,
+                channel_listing=variant_channel_listing,
+                rules_info=rules_info,
             )
         )
+
+    attach_voucher_info(lines_info, order)
+
+    return lines_info
+
+
+def attach_voucher_info(lines_info: list[EditableOrderLineInfo], order: Order):
+    """Collect necessary voucher info and attach it to order lines info."""
     voucher = order.voucher
     if voucher and (
         voucher.type == VoucherType.SPECIFIC_PRODUCT or voucher.apply_once_per_order
     ):
         voucher_info = fetch_voucher_info(voucher, order.voucher_code)
-        apply_voucher_to_line(voucher_info, lines_info)
-        denormalized_voucher_info = _get_denormalized_voucher_info(lines_info, voucher)
-        _apply_denormalized_voucher_to_line_info(
+        attach_voucher_to_line_info(voucher_info, lines_info)
+        denormalized_voucher_info = _fetch_denormalized_voucher_info(
+            lines_info, voucher
+        )
+        _attach_denormalized_voucher_to_line_info(
             lines_info, denormalized_voucher_info, order.voucher_code
         )
-    return lines_info
 
 
-def _get_denormalized_voucher_info(lines_info: list[EditableOrderLineInfo], voucher):
+def reattach_apply_once_per_order_voucher_info(
+    lines_info: list[EditableOrderLineInfo],
+    initial_cheapest_line_info: LineInfo | None,
+    order: Order,
+):
+    """Reattach apply once per order voucher info if the cheapest line has changed."""
+    if get_the_cheapest_line(lines_info) == initial_cheapest_line_info:
+        return
+
+    for line_info in lines_info:
+        line_info.voucher = None
+        line_info.voucher_code = None
+        line_info.voucher_denormalized_info = None
+
+    attach_voucher_info(lines_info, order)
+
+
+def _get_variant_listing(
+    variant: ProductVariant | None, channel_id: int
+) -> ProductVariantChannelListing | None:
+    if not variant:
+        return None
+    for channel_listing in variant.channel_listings.all():
+        if channel_listing.channel_id == channel_id:
+            return channel_listing
```

Comment:
This is to avoid additional db call (no use of `filter`), as channel_listings are prefetched